### PR TITLE
[highlighter] Add complex numbers pattern to our highlighter's `number` matching

### LINKED ISSUE #2214: [BUG] inconsistent coloring of complex numbers
Not sure if I should report this as a bug or as a feature request, but there is a very minor issue when printing complex numbers: the decimal part of the imaginary number is uncolored.

try for example:

```python
>>> from rich import pretty
>>> pretty.install()
>>> 3.05 + 2.06j
(3.05+2.06j)
```

image

I tried this on Linux, rich version 12.2.0. This "bug" appears both in the terminal (alacritty in my case) as well as in jupyter notebooks, which makes me believe it is platform independent.

### PR DESCRIPTION
**Pull request by DrBenton**
## Type of changes

- [x] Bug fix
- [ ] New feature
- [ ] Documentation / docstrings
- [ ] Tests
- [ ] Other

## Checklist

- [x] I've run the latest [black](https://github.com/psf/black) with default args on new code.
- [x] I've updated CHANGELOG.md and CONTRIBUTORS.md where appropriate.
- [x] I've added tests for new code.
- [x] I accept that @willmcgugan may be pedantic in the code review.

## Description

closes #2214 

The fix in action:

#### In IPython:
Screenshot from 2022-04-25 12-24-00

#### In a Jupyter notebook:
Screenshot from 2022-04-25 12-28-04

#### In plain Python 
Screenshot from 2022-04-25 12-29-03

### CODE REVIEW

**Review by willmcgugan — CHANGES_REQUESTED** [review]
I don't think this will work with negative imaginary numbers, e.g. `"1-3j"`

**Comment by willmcgugan** [comment]
No worries. We only really need to worry about what the repr for complex numbers can produce.

I'm afraid it does get more complex. If the imaginary part has an exponent, the regex fails.

e.g.  `(-123123-2.1312342342423422e+25j)`

I suspect that will occur almost never, but maybe we should try to support it.

I'm actually thinking that maybe the best approach would be to add a new regex for a new number type, which includes the parenthesis. So it matches `(<NUMBER>(+/-)<NUMBER>j)`. What do you think?

**Comment by DrBenton** [comment]
I made some progress, but still not there for the moment... :sweat_smile:  
Now my issue is that the "number" and "brace" patterns seem to take always precedence over the new "complex_number" one - which seem to make sense, at least for the brace, since `(<NUMBER>(+/-)<NUMBER>j)` does contain braces. Not sure what the best way to solve this could be? :thinking:

**Comment by DrBenton** [comment]
@willmcgugan I think I got there! It seems that I cannot prevent the "braces" to take precedence (even if I move the new `complex_number` matcher at the same level than it, outside of the `_combine_regex()` call), but apart from that it seems to work.

Here with a different visual style than "plain" numbers, just to illustrate that the matching is well and truly different:
Screenshot from 2022-04-25 16-44-08

What do you think? Good enough, or should we try harder to include the braces into this representation? :slightly_smiling_face:

**Comment by DrBenton** [comment]
Also, I just discovered that `3.4j` is a valid imaginary number, and in that case (only an imaginary part) there are no braces in its repersentation! 
Screenshot from 2022-04-25 17-03-42

File: rich/highlighter.py [hunk 1]
```diff
@@ -94,6 +94,7 @@ class ReprHighlighter(RegexHighlighter):
             r"(?P<call>[\w.]*?)\(",
             r"\b(?P<bool_true>True)\b|\b(?P<bool_false>False)\b|\b(?P<none>None)\b",
             r"(?P<ellipsis>\.\.\.)",
+            r"(?P<complex_number>(?<!\w)(?:\-?[0-9]+\.?[0-9]*(?:e[-+]?\d+?)?|0x[0-9a-fA-F]*)(?:[-+](?:[0-9]+\.?[0-9]*(?:e[-+]?\d+)?|0x[0-9a-fA-F]*))?j)",
```
Comment (willmcgugan) [inline rich/highlighter.py:97]: We might be able to simplify the regex. What we're formatting is the output of repr strings, which doesn't produce hex (that I'm aware of), so I think we can drop the `0x...` part.

**Review by willmcgugan — CHANGES_REQUESTED** [review]

File: rich/default_styles.py [hunk 2]
```diff
@@ -78,6 +78,7 @@
     "repr.attrib_equal": Style(bold=True),
     "repr.attrib_value": Style(color="magenta", italic=False),
     "repr.number": Style(color="cyan", bold=True, italic=False),
+    "repr.complex_number": Style(color="magenta", bold=True, italic=False),
```
[the remark this answers is not in the document]
Comment (DrBenton) [reply rich/default_styles.py]: Not sure we can, without adding complexity to the current machinery? The style is the concatenation of the ReprHighlighter's `base_style` ("repr.") and the regex pattern name - and it seems that we cannot use a dot in such a pattern name
> Group names must be valid Python identifiers, and each group name must be defined only once within a regular expression. 

We can have something close to it though: `repr.number_complex` - done in 69003a772c :slightly_smiling_face:

File: rich/highlighter.py (hunk 1 above)
Comment (DrBenton) [reply rich/highlighter.py]: done! (in 69003a772c)

File: rich/default_styles.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (willmcgugan) [reply rich/default_styles.py]: Ah. Fair enough.

**Review by willmcgugan — CHANGES_REQUESTED** [review]
Would you mind adding an entry in CHANGELOG.md with a link to the issue

**Comment by DrBenton** [comment]
> Would you mind adding an entry in CHANGELOG.md with a link to the issue

Sure! Done in [a02e5d6](https://github.com/Textualize/rich/pull/2219/commits/a02e5d675227c3cb6a60c77844ee92ca31ee745f) :slightly_smiling_face:

**Review by willmcgugan — APPROVED** [review]

**Pull request merged**